relayer/pkg/chainlink/keys/keystore.go

```diff
@@ -22,4 +126,57 @@ type NonceManager interface {
 
 	NextSequence(address caigotypes.Hash, chainID string) (*big.Int, error)
 	IncrementNextSequence(address caigotypes.Hash, chainID string, currentNonce *big.Int) error
-}
\ No newline at end of file
+}
+
+const (
+	maxPointByteLen = 32 // stark curve max is 252 bits
+	signatureLen    = 2 * maxPointByteLen
+)
+
+// signature is an intermediate representation for translating between a raw-bytes signature and a caigo
+// signature comprised of big.Ints
+type signature struct {
```

Comment:
Looking forward towards build separation (or even just our current blocking test entanglement), I think this type will be problematic because it can't be shared from both sides. We would need to have it represented in `chainlink-relay` to do that. Is it important for this to be a custom encoding? Or could we reuse the existing protobuf that already supports the reporting interfaces? https://github.com/smartcontractkit/chainlink-relay/blob/main/pkg/loop/internal/pb/relayer.proto#L306

Reply:
this is private and isn't shared. 

it's helper layer that is used in a specific implementation of the keystore adaption logic. in principle the keystore adapter is an interface and another implementation could be used, but there are a lot of constraints that have to be met, and this is private way to enforce those constraints for a specific implementation.

the protobuf definition doesn't help here because we need an encoding that turns bytes into multiple big.Ints. we would change the problem into needing an encoding for multiple protobuf instances

Reply:
more generally, we need a good story for this kind of pattern.

there is no constraint on new chains that they have to accept bytes as a signature and therefore there will be a need to adapt the loop.keystore interface to chain-specific logic

Reply:
> this is private and isn't shared, unless i'm misunderstanding you.

We need a way to understand the encoding on both ends. It would be better not to have separate redundant/duplicate code that needs to be kept in sync.

> we would change the problem into needing an encoding for multiple protobuf instances

Yes, but that is a trivial problem.

> more generally, we need a good story for this kind of pattern.
>
> there is no constraint on new chains that they have to accept bytes as a signature and therefore there will be a need to adapt the loop.keystore interface to chain-specific logic

This is why I suggested using a dynamic protobuf message instead of raw `[]byte`.

Reply:
I think https://protobuf.dev/programming-guides/proto3/#oneof is probably the idiomatic way to support different types of signatures here :shrug:

Reply:
>We need a way to understand the encoding on both ends. It would be better not to have separate redundant/duplicate code that needs to be kept in sync.

I don't think the explicit encoding needs to be exposed, rather we need a consistent use of the implemented interfaces.

the encoding is hidden behind the interfaces. if we use consistent implementations on both sides then only the implementors/owners of the chain need to think about this problem.

i dislike changing the protobuf defintion because it gets into a gray area of leaking the implementation details about a specific chains.

if a 3rd chain comes along with some other signature, we would need to update the proto rather to support it. that is a lot of overhead and coupling. instead we can leave it to the implementors to encode and decode the bytes as they need for there use case

it feels like the same kind of problem as marshaling and unmarshaling json or http -- as an implementor you need to know how to interpret the bytes, but the protocol that transports them doesn't care

Reply:
> if we use consistent implementations on both sides then only the implementors/owners of the chain need to think about this problem.

The implementation must exist in chainlink-relay, or be duplicated.

> i dislike changing the protobuf defintion because it gets into a gray area of leaking the implementation details about a specific chains.

The details are already leaked. This is just a way of standardizing them.

> if a 3rd chain comes along with some other signature, we would need to update the proto rather to support it. that is a lot of overhead and coupling. instead we can leave it to the implementors to encode and decode the bytes as they need for there use case

Core already needs to be updated, so a new relay version is not much more IMHO.

> it feels like the same kind of problem as marshaling and unmarshaling json or http -- as an implementor you need to know how to interpret the bytes, but the protocol that transports them doesn't care

The protocol is not the issue here. Core cares and it cannot import the chainlink-starknet.

Reply:
updated to use common code from the chainlink-relay repo. please review again